### DISCUSSION TITLE
How to debug Gradle + Groovy in VSCode?

### DISCUSSION THREAD

**Discussion by aentwist** [micronaut-projects/micronaut-core#7474] [Q&A]
I have a Gradle + Groovy Micronaut project that I want to debug in VSCode. The project has one caveat, it has a dotenv environment variable file .env.local.

I am currently running the project with `dotenv -c -- ./gradlew run`. To debug it, I add the `--debug-jvm` option. I am able to attach on port 5005.

I make a request that runs over a breakpoint, but execution does not stop at the breakpoint. Am I doing something wrong?

*Attach configuration:*

```json
{
  "type": "java",
  "request": "attach",
  "name": "Attach to Java App",
  "hostName": "localhost",
  "port": 5005
}
```

Also, if there is any way to start the project with a launch configuration, that would be better than starting it elsewhere and then attaching? My current launch configuration seems to hang. Maybe I need to use `preLaunchTask` somehow?

*Launch configuration:*

```json
{
  "type": "java",
  "request": "launch",
  "name": "Launch Java App",
  "mainClass": "com.example.Main",
  "envFile": "${workspaceFolder}/.env.local",
  "vmArgs": ["-Dorg.gradle.debug"]
}
```

**Comment by graemerocher** [comment]
This works out of the box if you uninstall VSCode for Java and install the "GraalVM Tools for Micronaut". Just open the `Main` class and hit F5 to debug.

See https://marketplace.visualstudio.com/items?itemName=oracle-labs-graalvm.micronaut
